### ISSUE TITLE
type-inference-independent comprehensions

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#7258]
https://groups.google.com/forum/#!topic/julia-dev/DF8LUiH7FP4

I was looking for an issue for this but realized that it wasn't a GitHub issue yet. Now it is.

**Comment by JeffBezanson** [comment]
The biggest problem here is empty ranges. Same problem as `map` of course. All we can do is return an `Array{Union()}` or throw an error.

**Comment by johnmyleswhite** [comment]
What do you think of throwing an error until we have a coherent idea what "should" happen? Seems relatively easy to make this not-prohibited behavior in the future.

**Comment by StefanKarpinski** [comment]
Throwing an error for mapping / comprehending over empty collections seems really broken. I think the only viable option is really to return `Array{Union()}`. It's unfortunate that we can return an `Any` array, but that would be a disaster for type stability. This means that `[x for x in [t...]]` is equivalent to `[t...]` for all tuple *except* the empty tuple. The old behavior where `[]` constructed `Array{Union()}(0)` was more consistent with the proposed behavior here.

**Comment by johnmyleswhite** [comment]
My problem with adopting the habit of using `Union()` in more places is that it forces users to try to understand type inference -- you have to learn that type-stability is important, except when taking a union operation with `Union()` being one of the operands.

**Comment by nalimilan** [comment]
Maybe the problem with `Array{Union()}` can be fixed elsewhere? I mean: what's frustrating with `Union()[]` is that you cannot do much with it, in particular `push!` -- which is probably one of the very few things people attempt and that doesn't work with `Union()[]` but does with `Any[]`. If we allowed `push!` on it (changing the result to `Array{Any}`), maybe returning `Union()[]` wouldn' t be an issue?

@johnmyleswhite has a point, but I'm not sure that's avoidable (cf. `Nullable{Union()}`).

**Comment by StefanKarpinski** [comment]
The whole point of this change is to make type inference no longer visible to the user since it doesn't affect the behavior of programs (just their performance).

**Comment by JeffBezanson** [comment]
@nalimilan You can't change the type of an existing value. Also I don't think using `push!` on the result of a comprehension is an important case. That just doesn't happen very often. It would make more sense to me to go the other direction, and have comprehensions return immutable arrays. After all part of the point of comprehensions is to have a fairly flexible way to initialize arrays without explicit assignment.

**Comment by nalimilan** [comment]
@JeffBezanson Yeah, special-casing `Union()[]` is probably too hackish. But always returning immutable arrays from comprehensions sounds a bit extreme. We could go with returning `Union()[]` and see if that works fine in practice.

**Comment by StefanKarpinski** [comment]
I really think that having `[x for x=[]]` return an array with element type `Union()` is the only viable option. Unfortunately, that disagrees with having `[]` produce an array with element type `Any`, but I guess that's probably ok.

**Comment by carnaval** [comment]
Is this the consensus ? I entirely agree about returning Vector{Bottom} for empty comprehension and `map` (which for now has the annoying behavior of defaulting to the input array type which makes absolutely no sense and bit me hard today), as well as to have them return immutable-length array to avoid confusion on the `push!` behavior.
One option is to have the typed version return mutable array, this way if you want to push on a comprehension result you just have to annotate the type. It's a minor annoyance and makes us completely independent of inference _and_ consistent which is great.
Another thing is that if we get `map` and comprehension to converge in behavior we could actually desugar the latter into the former once `map` is fast enough, which I think makes it a lot clearer for users (they can actually look at the implementation).
This is breaking so of course, the sooner the better etc etc

**Comment by Mike43110** [comment]
I do have a few cases where I use `push!` on a comprehension. This can be changed in my code but I just wanted to point out that there is at least one user that does use this. Mutable arrays would be preferable if at all possible.

**Comment by StefanKarpinski** [comment]
> One option is to have the typed version return mutable array, this way if you want to push on a comprehension result you just have to annotate the type. It's a minor annoyance and makes us completely independent of inference and consistent which is great.

100% agree with this approach. Do we have non-resizeable vectors? I know shared-data vectors can't be resized, but do we have the internal flags to mark that a vector shouldn't be resizeable even though it doesn't share data?

**Comment by StefanKarpinski** [comment]
> I do have a few cases where I use `push!` on a comprehension. This can be changed in my code but I just wanted to point out that there is at least one user that does use this. Mutable arrays would be preferable if at all possible.

@Mike43110, the alternative would be to return a mutable vector from `map` and thus allow you to push to the result without explicitly providing the return type (so far so good) BUT you'd have a corner case in your code where if the vector you're mapping over happens to be empty, the code will break. I think it's preferable to always require providing an element type for the result of `map` in order be able to push to the result array in exchange for not having your code be accidentally brittle to corner cases. The main place I can see that being a pain would be in the REPL where providing an element type is really annoying. But I can't really see any way around this sort of issue.

**Comment by Mike43110** [comment]
@StefanKarpinski I just was hoping for possible immutability. Using @carnaval's suggestion of allowing `push!` on explicitly typed comprehension/maps seems sound if a bit confusing without an explanation (i.e. why does typing allow for the array to be mutable now).

I use comprehensions to create arrays of initialised objects of a specific type, I don't think there is an alternative syntax for that. The proposed solution would be an easy change in my code where I can just annotate the types.

Erroring on empty inputs would be a bit of a painful change as I can think of cases where it would be valid code that would now require a check to see if a number is 0 before using it.

Ensuring the array you are pushing onto is correctly typed is a good thing anyway! Your string example in one of the 1st posts is also solved by making the Type `String` to allow for multiple types of strings.

Would a fixed length array (assuming #7568 is completed when this is implemented) not be preferable though to immutable? Or have I misunderstood something terribly wrong!

**Comment by carnaval** [comment]
I feel there is indeed a misunderstanding there. I think everyone saying immutable in this thread was actually using it as a shortcut for fixed length. Here is the reasoning (using map, but this is exactly the same for comprehension) :
- Explicitely typed is easy, say `map(::Function,ReturnType::Type,::Vector{InputType})::Vector{ReturnType}` which can be push!ed on as usual.
- Implicitly typed has the following problem : without executing the function, the only way to guess the return type is using inference, which is brittle because the semantic of the program does not depend on a syntactical construct, but on heuristics inside the compiler which can be non-obvious to the user. Of course, we can't execute the function even once if the input array is empty, we then have to choose a type. There are only two logical answers here (the two special points in the type lattice) :

You can choose Any, but you are then exposed to the non-obvious performance problem of suddenly boxing stuff because you happened to use a value from an empty call to map earlier in the program.

Or, you can choose Bottom, which has the nice following property : if the return type of map is `Union(Vector{ReturnType},Vector{Bottom})` then the inference algorithm can eliminate the second component in traces where you actually use the array, for example by indexing in it. Much in the same way that the return type of `throw()` is Bottom, which allows us to know the type of `f(throw(something))` (Bottom again) without knowing f at all.

The problem with returning the empty Bottom array, is that you can't push! in it. It would be again a strange pitfall to be able to push! inside the result of map in most cases except when it is empty, which brings us to the only sane way of being consistent : disallow mutation of the _length_ of the result of map when the return type was not explicit.

Does this makes sense ?

**Comment by Mike43110** [comment]
It makes perfect sense thanks! 

Then by specifying the return type, a fixed length array can anyway be selected as the output if I am not mistaken?

Another nice thing with fixed length arrays when untyped maps/comprehensions are done is that it is less likely that somebody will try and put an incorrect type into the array. Barring `setindex!` of course.

Returning `Any` really does seem like a terrible idea! One could not expect it as the length could be a result from random where a 0 length array would make sense. Using `push!` on this would keep the array as `Any` for that run and give a performance hit. This would be difficult to detect as it may not appear in all runs.

Thanks for fixing my misunderstanding!

**Comment by toivoh** [comment]
To be safe from `setindex!` as well, the returned array would actually have to be immutable. I could definitely see a point to that, if it weren't for the fact that we don't have any immutable arrays.

**Comment by simonster** [comment]
Besides `push!`, the other problem case that I don't see discussed here is:

```julia
f(::Vector{Int}) = true
g(x::Vector{Int}) = f([y+1 for y in x])
g(Int[])
```

Unfortunately I don't think there is a way around this. This is a further argument for using `Union()` as the element type in that at least:

```julia
f{T<:Integer}(::Vector{T}) = true
g{T<:Integer}(x::Vector{T}) = f([y+1 for y in x])
g(Int[])
```

would still work.

**Comment by yuyichao** [comment]
Feels like this can easily introducing ambiguity that has to be handled.

i.e. what if the orginal design was sth like
```julia
f(::Vector{Int}) = true
f(::Vector{FloatingPoint}) = false
```

**Comment by carnaval** [comment]
Well there is no good answer then. What does `f([rand(Bool) ? 1.0 : 1 for _ in []])` should be ?

**Comment by JeffBezanson** [comment]
I'm perfectly comfortable judging the behavior based on the length of its description (where quality = 1/length), and "the element type of the result is the join of the types of the elements" passes the test IMO.

**Comment by carnaval** [comment]
By this metric I think the best solution is "there is no empty array". Or maybe "what empty array ?".

**Comment by carnaval** [comment]
@JeffBezanson I know you have a pile of more urgent issues to go through right now, but ...

Is there any way we could get this feature (i.e., make `[x for y in z]` lower to an inlined version of `map(y->x,z)`) on a branch/under a flag or something ? It would allow for breakage testing and to be honest I kinda need it because the green fairy is punting on type_goto and those are hard to avoid in Base code.

If you don't see yourself doing this soon please tell, I'll try to hack something in the frontend but I'm pretty sure it would be a lot easier if you did it ;)

**Comment by carnaval** [comment]
I think we have to move this one out of the 0.4 milestone unfortunately. If we go with the typejoin implementation at least, because of the possible widespread perf regressions we would have to track down. I don't think it's realistic to block the release until then.

**Comment by yuyichao** [comment]
I was going to ask if we can keep the current inference code and narrow it down at runtime if necessary.

**Comment by carnaval** [comment]
I'm not sure I get what you mean. Inference tries to narrow down the types of any julia code, so in my implementation, if inference figures out the type of the expression, the slow path should be removed. The main problem is the performance issue of having every untyped comprehension return `::Union(Vector{T},Vector{Union()})`. Coupled with the fact that we do union widening with a hammer for now (see the old TODO in inference) I'm quite sure that many places will end up slow (or maybe incorrect if they try to dispatch on the array eltype but that's another issue).

**Comment by yuyichao** [comment]
I see. I guess I didn't know about the fast path.

**Comment by StefanKarpinski** [comment]
Counterargument: we could do this for 0.4 and suffer the performance issues until we manage to fix performance in 0.4.1 or so. I know that's not optimal but performance regressions are not entirely show-stoppers for releases, especially when there do exist fairly straightforward workarounds, like explicitly specifying the element type.

**Comment by carnaval** [comment]
I think there are more urgent things to fix than add another breaking change + probable perf regression a few weeks (wishful thinking ?) before release. I'll move this to 0.4.X.

**Comment by yuyichao** [comment]
Since this is breaking, shouldn't it be 0.5.0 than?

**Comment by carnaval** [comment]
Oh sure, I'm just trying to remove things from 0.4 :)

**Comment by juliohm** [comment]
What is the status of this issue?

**Comment by StefanKarpinski** [comment]
Just waiting on (re)implementation – @carnaval make a PR that could be rebased.

**Comment by JeffBezanson** [comment]
Was waiting for #15300 first. We probably also want #15146 first, plus performance improvements (#15379 might help).

**Comment by tkelman** [comment]
Blocked on performance regressions at the moment? Still has a branch that could be rebased (significantly) for testing.

**Comment by StefanKarpinski** [comment]
Yes, all of the blocking issues/prs are now merged, so we should revisit whether this can be done now.

**Comment by JeffBezanson** [comment]
I think #15402 is the relevant regression here.

**Comment by JeffBezanson** [comment]
Current PR: #16622
Blocked by: #16201 
which is fixed by: #16692

**Comment by StefanKarpinski** [comment]
Status update?

**Comment by JeffBezanson** [comment]
Status is same as last week.
